distributed/tests/test_worker.py

```diff
@@ -899,27 +900,103 @@ def __sizeof__(self):
     assert result.data == 123
 
 
+class Bad:
+    def __init__(self, size):
+        self.size = size
+
+    def __getstate__(self):
+        raise TypeError()
+
+    def __sizeof__(self):
+        return self.size
+
+
 @gen_cluster(client=True)
 async def test_fail_write_to_disk(c, s, a, b):
-    class Bad:
-        def __getstate__(self):
-            raise TypeError()
-
-        def __sizeof__(self):
-            return int(100e9)
 
-    future = c.submit(Bad)
+    """Test failure to spill triggered by key which is individually larger
+    than target. The data is lost and the task is marked as failed;
+    the worker remains in useable condition.
+    """
+    future = c.submit(Bad, int(100e9))
     await wait(future)
 
     assert future.status == "error"
 
-    with pytest.raises(TypeError):
+    with pytest.raises(TypeError, match="Could not serialize"):
         await future
 
     futures = c.map(inc, range(10))
-    results = await c._gather(futures)
+    results = await c.gather(futures)
+    assert results == list(map(inc, range(10)))
+
+
+@gen_cluster(
+    client=True, 
+    nthreads=[("", 1)],
+    worker_kwargs=dict(
+        memory_limit=1000,
+        memory_target_fraction=0.5,
+        memory_spill_fraction=False,
+        memory_pause_fraction=False,
+    )
+)
+async def test_fail_write_to_disk_2(c, s, a):
+    """Test failure to spill triggered by key which is individually smaller
+    than target, so it is not spilled immediately. The data is retained and
+    the task is NOT marked as failed; the worker remains in useable condition.
+    """
+    x = c.submit(Bad, 250)
+    await wait(x, key="x")
+    assert x.status == "memory"
+    assert set(a.data.memory) == {"x"}
+    
+    
+    y = c.submit(lambda: "y" * 500, key="y")
+    await(y)
+    assert set(a.data.memory) == {"x", "y"}
+    assert not a.data.disk
+
+    futures = c.map(inc, range(10))
+    results = await c.gather(futures)
     assert results == list(map(inc, range(10)))
 
+try:
+    import zict
+except ImportError:
+    zict = None
+requires_zict_210 = pytest.mark.skipif(
+    not zict or parse_version(zict.__version__) <= parse_version("2.0.0"),
+    reason="requires zict version > 2.0.0",
+)
+
+
+@requires_zict_210
+@gen_cluster(
+    client=True,
+    nthreads=[("", 1)],
+    worker_kwargs=dict(
+        memory_limit=1,  # Spill everything
+        memory_target_fraction=False,
+        memory_spill_fraction=0.01,
+        memory_pause_fraction=False,
+    ),
+)
+async def test_fail_write_to_disk_evict(c, s, a):
+    """Test failure to evict a key, triggered by the spill threshold"""
+
+    with captured_logger(logging.getLogger("distributed.spill")) as logs:
+        bad = c.submit(Bad, 1, key="bad")
+        await wait(bad)
+
+    logs_value = logs.getvalue()
+    assert "Failed to pickle" in logs_value
```

Comment:
@crusaderky I modify the test and included your suggestion of pulling the "distributed.spill" logs instead of "distributed.protocol.pickle" logs, but this causes the test to fail since the logs are empty. Are we pulling the right logs? or the problem is somewhere else?